crates/matrix-sdk-indexeddb/src/stream.rs

```diff
@@ -0,0 +1,624 @@
+                let transaction = (this.transaction_builder)()?;
+                this.latest_transaction.set(transaction);
+
+                // Get and asve the new `object_store`.
+                let object_store = unsafe { this.latest_transaction_ptr.as_ref() }
+                    .object_store(this.object_store_name.as_str())?;
+                this.latest_object_store.set(Some(object_store));
+
+                // Get a new `cursor_future`.
+                let object_store_ref = unsafe { this.latest_object_store_ptr.as_ref() };
+                let after_latest_key = IdbKeyRange::lower_bound_with_open(&this.latest_key, true)?;
+                let cursor_future =
+                    object_store_ref.as_ref().unwrap().open_cursor_with_range(&after_latest_key)?;
+
+                // Save the `cursor_future` to poll it later.
+                *this.cursor_future = Some(Box::pin(cursor_future));
+            }
+
+            // Let's poll `cursor_future` since it's `Some(_)`!
+            let cursor = ready!(this
+                .cursor_future
+                .as_mut()
+                // SAFETY: The case where `cursor_future` is `None` has been handled
+                // hereinabove.
+                .unwrap()
+                .as_mut()
+                .poll(cx))?;
+
+            // We have a cursor. Yeah! We can build the `StreamByCursor`.
+            let inner_stream = StreamByCursor::new_with_cursor(cursor);
+            // And save it.
+            *this.inner_stream = Some(inner_stream);
+
+            // Update `renew_in`.
+            *this.renew_in = *this.renew_every;
+        }
+
+        // Let's poll `inner_stream`.
+        let result = ready!(this
+            .inner_stream
+            .as_pin_mut()
+            .expect("`inner_stream` cannot be `None`")
+            .poll_next(cx));
+
+        if let Some(Ok((key, _value))) = &result {
+            // Decrement `renew_in` only if a result has been produced.
+            *this.renew_in -= 1;
+            // Save the latest key.
+            *this.latest_key = key.clone();
+        }
+
+        Poll::Ready(result)
+    }
+}
+
+#[cfg(all(test, target_arch = "wasm32"))]
+mod tests {
+    use futures_util::TryStreamExt;
+    use indexed_db_futures::prelude::*;
+    use wasm_bindgen::JsValue;
+    use wasm_bindgen_test::{wasm_bindgen_test, wasm_bindgen_test_configure};
+    use web_sys::DomException;
+
+    use super::*;
+
+    wasm_bindgen_test_configure!(run_in_browser);
+
+    #[wasm_bindgen_test]
+    async fn stream_by_cursor() -> Result<(), DomException> {
+        let mut db = IdbDatabase::open("foobar")?;
+
+        db.set_on_upgrade_needed(Some(|event: &IdbVersionChangeEvent| -> Result<(), JsValue> {
+            if event.db().object_store_names().find(|n| n == "baz").is_none() {
+                event.db().create_object_store("baz")?;
+            }
+
+            Ok(())
+        }));
+
+        let db = db.await?;
+
+        {
```

Comment:
It might be nice to put this into a helper function to shorten the test and make it easier to understand.